### PR TITLE
add module and state to help create txt docs from highstate data

### PR DESCRIPTION
**Pull request by steverweber**
### What does this PR do?

auto generates system documents in markdown format using highstate information.

https://github.com/saltstack/salt/pull/41920#issuecomment-311768276

### CODE REVIEW

File: salt/modules/highstate_doc.py [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+    salt-call state.highstate
+    salt-call state.apply makereadme
+    # or if you dont want the extra `make helpfile` state
+    salt-call --out=newline_values_only salt.highstate_doc.render > /root/README.md ; chmod 0600 /root/README.md
+
+
+Creating a document collection
+------------------------------
+
+From the master we can run the following Script to
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: `script` not `Script`

File: salt/modules/highstate_doc.py [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+            # mkdir m
+            #directory = os.path.dirname(file_path)
+            if not os.path.exists(m):
+                os.makedirs(m)
+            with open(m + '/README.md','wb') as f:
+                f.write(d)
+            print('ADDED: ' + m + '/README.md')
+
+
+Once the master has a collection of all the readmes.
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: `README files` instead of `readmes`

File: salt/modules/highstate_doc.py [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+You can use pandoc to create html versions of the markdown.
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: `HTML`

File: salt/modules/highstate_doc.py [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+        ff=${f#$dest/}
+        minion=${ff%%/*}
+        echo "proccess: $dest/${minion}/$(basename $f)"
+        cat $dest/${minion}/$(basename $f) | \
+            pandoc --standalone --from markdown_github --to html \
+            --include-in-header $dest/style.html \
+            > $dest/${minion}/$(basename $f).html
+    done
+
+Its also nice to put the help files in source control.
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: `It is`

File: salt/modules/highstate_doc.py [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+num_cpus: {{grains.get('num_cpus')}}
+ipv4: {{grains.get('ipv4')}}
+master: {{opts.get('master')}}
+####################################################
+```
+
+This system is fully or partly managed using Salt.
+
+The following sections are a rendered view of what the configuration management system
+manages on this system. Each item is handled in order from top to bottom unless some
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: Maybe `controlled` instead of `manages`?

File: salt/modules/highstate_doc.py [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+                with open(m + '/README.md','wb') as fin:
+                    fin.write(d)
+                print('ADDED: ' + m + '/README.md')
+    '''
+    out = ''
+    try:
+        with salt.utils.fopen(name, 'r') as f:
+            out = f.read()
+    except Exception as ex:
+        return None
```
Comment (cachedout) [inline salt/modules/highstate_doc.py]: Do we want to at least log an error if the read fails?

Comment (steverweber) [inline salt/modules/highstate_doc.py]: sure I'll add that.

File: salt/modules/highstate_doc.py (hunk 1 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

File: salt/modules/highstate_doc.py (hunk 2 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

File: salt/modules/highstate_doc.py (hunk 3 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

File: salt/modules/highstate_doc.py (hunk 4 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

File: salt/modules/highstate_doc.py (hunk 5 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

File: salt/modules/highstate_doc.py (hunk 6 above)
Comment (steverweber) [inline salt/modules/highstate_doc.py]: done

**Review by cachedout — APPROVED** [review]

**Pull request merged**